# Multi-genotype analyses of long-ncRNAs in Sugarcane

## Identification of long non-coding RNAs in sugarcane
* [Predicting putative lncRNAs with CPC2](https://github.com/labbces/sugarcane_RNAome/wiki/Predicting-putative-lncRNAs-with-CPC2)
* [Predicting putative lncRNAs with PLncPRO](https://github.com/labbces/sugarcane_RNAome/wiki/Predicting-putative-lncRNAs-with-PLncPRO)
* [Predicting putative lncRNAs with RNAplonc](https://github.com/labbces/sugarcane_RNAome/wiki/Predicting-putative-lncRNAs-with-RNAplonc)
* [Aggregating the consensus of putative ncRNAs predicted by the 3 tools](https://github.com/labbces/sugarcane_RNAome/wiki/Aggregating-the-consensus-of-putative-ncRNAs-predicted-by-the-3-tools)

The following results represent the percentage of sequences classified as lncRNAs from the 3 softwares used in identifying putative lncRNAs in sugarcane (described in detail above). The `16,237,098 input sequences` originate from `48 sugarcane transcriptomes` that were assembled and analyzed in a previous project, where we conducted inference and annotation of the [sugarcane pan-transcriptome](https://github.com/labbces/SCPT).

![predicting_lncRNAs_results](https://github.com/labbces/sugarcane_RNAome/blob/main/images/predicting_lncRNAs_results.PNG)

## Conserved long non-coding RNAs in the panRNAome
* [Clustering sequences by identity](https://github.com/labbces/sugarcane_RNAome/wiki/Clustering-sequences-by-identity)
* [Inference of the pan-ncRNAome](https://github.com/labbces/sugarcane_RNAome/wiki/Inference-of-the-pan%E2%80%90ncRNAome)

![panRNAomeGroupsTrajectory_0.8.png](https://github.com/labbces/sugarcane_RNAome/blob/main/scripts/panTrajectory/panRNAome/panRNAomeGroupsTrajectory_0.8.png)

## Families/classes of lncrnas
* [Annotation of ncRNAs families](https://github.com/labbces/sugarcane_RNAome/wiki/Annotation-of-ncRNAs-families)

![infernal_cpclncrnas_families_48_normalized_annotation.png](https://github.com/labbces/sugarcane_RNAome/blob/main/scripts/runInfernal/panRNAomeRfamFamilies.png)

## Co-expression patterns of lncRNAs
* [Contrasting genotypes selection](https://github.com/labbces/sugarcane_RNAome/wiki/Contrasting-genotypes-selection)
* [Calculating the RNA expression for each sample](https://github.com/labbces/sugarcane_RNAome/wiki/Calculating-the-RNA-expression-for-each-sample)
* [Filtering the RNAs expression matrix of the contrasting genotypes](https://github.com/labbces/sugarcane_RNAome/wiki/Filtering-the-RNAs-expression-matrix-of-the-contrasting-genotypes)
* [Filtering the RNAs expression matrix by function](https://github.com/labbces/sugarcane_RNAome/wiki/Filtering-the-RNAs-expression-matrix-by-function)
* [Analyzing RNA expression patterns after filtering](https://github.com/labbces/sugarcane_RNAome/wiki/Analyzing-RNA-expression-patterns-after-filtering)
* [Principal Component Analysis of lncRNAs expression in contrasting genotypes](https://github.com/labbces/sugarcane_RNAome/wiki/Principal-Component-Analysis-of-lncRNAs-expression-in-contrasting-genotypes)
* [Computing gene pair relationships with Pearson correlation](https://github.com/labbces/sugarcane_RNAome/wiki/Computing-gene-pair-relationships-with-Pearson-correlation)
* [Analyzing MCL clusterings (effects of Inflation value)](https://github.com/labbces/sugarcane_RNAome/wiki/Analyzing-MCL-clusterings-(effects-of-Inflation-value))
* [Finding conserved modules of co-expression (complete subgraphs - cliques)](https://github.com/labbces/sugarcane_RNAome/wiki/Finding-conserved-modules-of-co%E2%80%90expression-(complete-subgraphs-%E2%80%90-cliques))
* [Visualizing co-expressed gene modules in contrasting genotypes](https://github.com/labbces/sugarcane_RNAome/wiki/Visualizing-co%E2%80%90expressed-gene-modules-in-contrasting-genotypes)
* [Gene Ontology enrichment analysis: exploring biological function of modules](https://github.com/labbces/sugarcane_RNAome/wiki/Gene-Ontology-enrichment-analysis:-exploring-biological-function-of-modules)
